### ISSUE TITLE
Write stream of unordered rows into partitioned table causes "Already closed files for partition"

### ISSUE THREAD

**Issue by living42** [apache/iceberg#717]
I'am trying to use Structured Streaming to move data from Kafka to Iceberg table. some time, the data is out of order according to the partition spec, then i got a exception like below:

```
java.lang.IllegalStateException: Already closed files for partition: time_day=2019-12-17
	at org.apache.iceberg.spark.source.Writer$PartitionedWriter.write(Writer.java:505)
	at org.apache.iceberg.spark.source.Writer$PartitionedWriter.write(Writer.java:476)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$$anonfun$run$3.apply(WriteToDataSourceV2Exec.scala:118)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$$anonfun$run$3.apply(WriteToDataSourceV2Exec.scala:116)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1394)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:146)
	at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec$$anonfun$doExecute$2.apply(WriteToDataSourceV2Exec.scala:67)
	at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec$$anonfun$doExecute$2.apply(WriteToDataSourceV2Exec.scala:66)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

I wish i can use `sortWithinPartitions` operation on streaming, but Spark won't let me do that.

**Comment by rdblue** [comment]
Yes, `sortWithinPartitions` won't work, but you should be able to `repartition` by your partition columns, right?

@aokolnychyi and I have also talked about relaxing this constraint for Spark streaming and keeping multiple files open.

**Comment by living42** [comment]
Yes, repartition works, but it feels like it’s a extra work to workaround the constraint.

**Comment by davseitsev** [comment]
We also got into this restriction when write data from Kafka to iceberg and partition it by date.
We would like to have a way avoid the restriction, write data as is and then compact small files in background.

**Comment by HeartSaVioR** [comment]
fanout writer will enable you to do this without repartition/sort. There's no doc yet (as I guess the functionality will be available in 0.11) so please refer https://github.com/apache/iceberg/pull/1929 to let yourself preview the functionality.
